telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala

```diff
@@ -0,0 +1,205 @@
+
+import scala.collection.JavaConverters._
+import org.apache.kafka.clients.producer.KafkaProducer
+import com.twitter.util.Duration
+import com.twitter.util.Awaitable.CanAwait
+import com.twitter.util.Time
+import com.twitter.util.Future
+import com.twitter.util.Try
+import com.google.common.collect.EvictingQueue
+import com.twitter.finagle.tracing.Tracer
+import java.util.concurrent.Semaphore
+
+import com.twitter.finagle.{Stack, tracing}
+import com.twitter.finagle.zipkin.thrift.ScribeRawZipkinTracer
+
+import com.twitter.conversions.time._
+import com.twitter.finagle.stats.StatsReceiver
+import com.twitter.finagle.tracing.Tracer
+import com.twitter.finagle.util.DefaultTimer
+import com.twitter.util._
+import java.net.InetSocketAddress
+import java.nio.ByteBuffer
+import com.twitter.finagle.zipkin.core.RawZipkinTracer
+import com.twitter.finagle.zipkin.core.Span
+import com.twitter.finagle.zipkin.core.DeadlineSpanMap
+import com.twitter.finagle.Codec
+import java.io.ByteArrayOutputStream
+import java.io.ObjectOutputStream
+import java.io.IOException
+import org.apache.kafka.clients.producer.ProducerRecord
+import org.apache.kafka.clients.producer.Callback
+import org.apache.kafka.clients.producer.RecordMetadata
+import org.apache.thrift.TSerializer
+import com.twitter.finagle.zipkin.core.TracerCache
+
+import java.util.Map
+import java.util.HashMap
+
+import org.apache.kafka.clients.producer.ProducerConfig
+import com.twitter.logging.Logger
+
+class KafkaTelemeterInitializer extends TelemeterInitializer {
+  type Config = KafkaTelemeterConfig
+  override def configId = "io.buoyant.telemetry.KafkaTelemeterConfig"
+  def configClass = classOf[KafkaTelemeterConfig]
+}
+
+case class KafkaTelemeterConfig(brokerList: String, sampleRate: Float, numRetries: Int, metrics: Boolean, tracing: Boolean) extends TelemeterConfig {
+  private val log = Logger.get(getClass)
+  def mk(params: Stack.Params): KafkaTelemeter = {
+    log.info("Broker list is %s", brokerList)
+    log.info("Number of retries per request is %s", brokerList)
+    new KafkaTelemeter("zipkin", numRetries, sampleRate, brokerList, metrics, tracing)
+  }
+}
+
+//Going with hybrid approach. Let kafka handle retries. Batching will be handled outside as it is being done today.
+//We will leave linger.ms unchanged for this reason at 1 ms.
+//The reason for this is that the API has been constructed in such a way that a given trace is annotated across multiple
+//different calls rather than gathering up all its annotations and then sending it on its way.
+
+case class KafkaTelemeter(topic: String, numRetries: Int, sampleRate: Float, brokerList: String, metrics: Boolean, tracing: Boolean) extends Telemeter {
+  private val log = Logger.get(getClass)
+  log.info("Initializing kafka telemeter")
+  private var producer: KafkaProducer[Array[Byte], Array[Byte]] = null
+  private val METADATA_FETCH_TIMEOUT_CONFIG: Int = 3000
+  private val TIMEOUT_CONFIG: Int = 300
```

Comment:
That is only in the newer versions of kafka. Again look at my previous comments.